app/scripts/controllers/network/network-controller.test.js

```diff
@@ -50,224 +69,4106 @@ const PRE_1559_BLOCK = {
+                  },
+                });
+
+                const promiseForInfuraIsUnblocked = waitForEvent({
+                  controller,
+                  eventName: 'infuraIsUnblocked',
+                  operation: async () => {
+                    await controller.initializeProvider();
+                  },
+                });
+
+                await expect(promiseForInfuraIsUnblocked).toNeverResolve();
+              },
+            );
+          });
+        });
+
+        it(`persists "${networkVersion}" to state as the network version of ${networkName}`, async () => {
+          await withController(
+            {
+              state: {
+                provider: {
+                  type: networkType,
+                },
+              },
+            },
+            async ({ controller, network }) => {
+              network.mockEssentialRpcCalls();
+              await withoutCallingLookupNetwork({
+                controller,
+                operation: async () => {
+                  await controller.initializeProvider();
+                },
+              });
+
+              await waitForStateChanges({
+                controller,
+                propertyPath: ['network'],
+                operation: async () => {
+                  await controller.lookupNetwork();
+                },
+              });
+
+              expect(controller.store.getState().network).toStrictEqual(
+                networkVersion,
+              );
+            },
+          );
+        });
+
+        it(`does not update the network state if it is already set to "${networkVersion}"`, async () => {
+          await withController(
+            {
+              state: {
+                provider: {
+                  type: networkType,
+                },
+              },
+            },
+            async ({ controller, network }) => {
+              network.mockEssentialRpcCalls({
+                eth_blockNumber: {
+                  times: 2,
+                },
+              });
+              await withoutCallingLookupNetwork({
+                controller,
+                operation: async () => {
+                  await controller.initializeProvider();
+                },
+              });
+
+              await waitForStateChanges({
+                controller,
+                propertyPath: ['network'],
+                operation: async () => {
+                  await controller.lookupNetwork();
+                },
+              });
+
+              const promiseForStateChanges = waitForStateChanges({
+                controller,
+                propertyPath: ['network'],
+                operation: async () => {
+                  await controller.lookupNetwork();
+                },
+              });
+
+              await expect(promiseForStateChanges).toNeverResolve();
+            },
+          );
+        });
+
+        describe('if the request for eth_getBlockByNumber responds successfully', () => {
```

Comment:
Nit: It would be great to cover the case where `null` is returned by a successful response as well. IIRC that is expected if the network doesn't recognize the requested block number.

Reply:
Ah. I missed that while reading the code. Good point. Added here: [520ad6d](https://github.com/MetaMask/metamask-extension/pull/17498/commits/520ad6dda775b3b9cd10e028e0fab8a12fbb5bb8)